sdks/java/core/src/main/java/org/apache/beam/sdk/coders/AvroCoder.java

```diff
@@ -321,38 +298,79 @@ public void verifyDeterministic() throws NonDeterministicException {
   @Deprecated
   public DatumWriter<T> createDatumWriter() {
     if (type.equals(GenericRecord.class)) {
-      return new GenericDatumWriter<>(schema);
+      return new GenericDatumWriter<>(getSchema());
     } else {
-      return new ReflectDatumWriter<>(schema);
+      return new ReflectDatumWriter<>(getSchema());
     }
   }
 
   /**
    * Returns the schema used by this coder.
    */
   public Schema getSchema() {
-    return schema;
+    return getMemoizedSchema();
+  }
+
+  /**
+   * Get the memoized {@link BinaryDecoder}, possibly initializing it lazily.
+   */
+  private ThreadLocal<BinaryDecoder> getDecoder() {
+    if (memoizedDecoder == null) {
```

Comment:
Do you mean the instantiation of memoizedDecoder should be synchronized so only one is created? I went by what was done in ProtoCoder, but synchronization can be added here.